### ISSUE TITLE
GLSL compute shader does not recompile when updated

### ISSUE THREAD

**Issue by mlepage** [godotengine/godot#72281]
### Godot version

4.0 beta 16

### System information

Windows 11

### Issue description

I follow the instructions as in the example
https://docs.godotengine.org/en/latest/tutorials/shaders/compute_shaders.html

and it works and outputs the values doubled.

I edit the shader so instead of doubling the values (*= 2.0) it triples the values (*= 3.0).

I run the program but the values are still doubled (as before), not tripled (as updated).

In fact, if I type garbage into the glsl file, it still works. It's not updating the compiled version of the shader.

If I rename the glsl resource file to .txt then back to .glsl, it recompiles the shader and then I see the updated values. But this renaming dance should not be necessary when updating the shader code.

### Steps to reproduce

https://docs.godotengine.org/en/latest/tutorials/shaders/compute_shaders.html

Then edit it to triple instead of double, and run it again.

### Minimal reproduction project

N/A

**Comment by clayjohn** [comment]
I think you need to reimport the GLSL file if it changes after you imported it originally.

**Comment by mlepage** [comment]
At one point I did see some kind of error in the output along the lines of "shader changed signal" so maybe it is somehow trying to update, but failing. I don't know enough about the pipeline to know how it is supposed to recompile.

I made the shader entirely in Godot 4 editor and am working on it with the corresponding gdscript at the same time. Is there an easy way to reimport the GLSL file?

**Comment by clayjohn** [comment]
> At one point I did see some kind of error in the output along the lines of "shader changed signal" so maybe it is somehow trying to update, but failing. I don't know enough about the pipeline to know how it is supposed to recompile.
> 
> I made the shader entirely in Godot 4 editor and am working on it with the corresponding gdscript at the same time. Is there an easy way to reimport the GLSL file?

Yes, you click on the file in the asset explorer, then click import > reimport, just like you would with any other resource

**Comment by mlepage** [comment]
OK I found the import tab and that does help. Should it auto import when the underlying GLSL file (which is in the project) is edited in the editor? Or is that not expected?

**Comment by clayjohn** [comment]
> OK I found the import tab and that does help. Should it auto import when the underlying GLSL file (which is in the project) is edited in the editor? Or is that not expected?

I don't think so. We don't have a built in editor for custom glsl files, they are meant to be edited in an external program. For Godot shaders we have a dedicated editor which allows the engine to automatically detect changes and recompile the shader, but we don't have anything like that for custom shaders.

**Comment by mlepage** [comment]
OK so this is working as expected and not a bug?

**Comment by Calinou** [comment]
> We don't have a built in editor for custom glsl files, they are meant to be edited in an external program

Could something similar to https://github.com/godotengine/godot/pull/72259 be done to allow editing GLSL files in the script editor?

**Comment by bitsawer** [comment]
@mlepage I would still consider this a bug. I think if the Godot editor supports a certain file type it should automatically re-import it if the file changes in the file-sytem, even by external programs (which should happen when the editor scans the filesystem on focus gain etc.). If the automatic re-import does not happen it sounds like a bug, unless I misunderstood something. Otherwise teams would have nightmarish source-control updates where they would have to go through every changed file and manually re-import them or risk using out-of-sync data.

**Comment by mlepage** [comment]
> > We don't have a built in editor for custom glsl files, they are meant to be edited in an external program
> 
> Could something similar to #72259 be done to allow editing GLSL files in the script editor (not the shader editor)?

Yes exactly, I've been editing them in the script editor.

**Comment by clayjohn** [comment]
I think right now editing in the script editor is kind of hacky and isn't a supported workflow. I agree it would be nice to edit them in the script editor and have Godot automatically reimport the shader when it changes. So I am marking this as an enhancement discussion

**Comment by Calinou** [comment]
> Yes exactly, I've been editing them in the script editor.

I'm not referring to using the current hacky approach, but doing something similar to https://github.com/godotengine/godot/pull/72259 so that a *resource* file can be edited properly in the script editor.

**Comment by Flarkk** [comment]
> > At one point I did see some kind of error in the output along the lines of "shader changed signal" so maybe it is somehow trying to update, but failing. I don't know enough about the pipeline to know how it is supposed to recompile.
> > I made the shader entirely in Godot 4 editor and am working on it with the corresponding gdscript at the same time. Is there an easy way to reimport the GLSL file?
> 
> Yes, you click on the file in the asset explorer, then click import > reimport, just like you would with any other resource

Is reimporting supposed to trigger the `changed` signal on the resource instances ?
I'm speaking here in the context of editor development, where I need to trigger some action (typically re-run some compute shaders to bake in textures) each time the user changes the GLSL shader code.
I saw `emit_changed()` is called especially in [`RDShaderFile::set_bytecode()`](https://github.com/godotengine/godot/blob/86bf8354a06ab7b23a0ff6a81b48fd015e92ac94/servers/rendering/rendering_device_binds.h#L340-L344) but from what I experimented so far the signal seem not to be triggered on reimport. Is `RDShaderFile::set_bytecode()` called at all when a GLSL file is reimported ?
Not sure how close the above is related to this specific ticket. Any view welcome.

**Comment by Shardj** [comment]
This has caused me so much frustration, thinking my compute shader code was broken and trying to fix bugs that don't exist. It's insane that Godot doesn't detect these changes and recompile the SPIR-V

Edit: in my case, it seemed to be an issue specific to glslinc file use, moving everything into a single .glsl file seems to be far safer and behaviour is consistent again

**Comment by Shardj** [comment]
For others having issues with changes to glslinc files not being noticed by the godot editor, here's a plugin to fix it https://github.com/Shardj/glslinc-reimport-fix as it's been driving me crazy
